### ISSUE TITLE
ld.lld: error: section exceeds available address space (MIPS vmlinuz.bin)

### ISSUE THREAD

**Issue by nathanchance** [ClangBuiltLinux/linux#1546]
When building `ARCH=mips gcw0_defconfig vmlinuz.bin`, I see the following error:

```
$ make -kj"$(nproc)" ARCH=mips LLVM=1 gcw0_defconfig vmlinuz.bin
...
  LD      vmlinux
  SORTTAB vmlinux
  SYSMAP  System.map
  AS      arch/mips/boot/compressed/head.o
  CC      arch/mips/boot/compressed/decompress.o
  CC      arch/mips/boot/compressed/string.o
  CC      arch/mips/boot/compressed/bswapsi.o
  CC      arch/mips/boot/compressed/dummy.o
  OBJCOPY arch/mips/boot/compressed/vmlinux.bin
  HOSTCC  arch/mips/boot/compressed/calc_vmlinuz_load_addr
  GZIP    arch/mips/boot/compressed/vmlinux.bin.z
  OBJCOPY arch/mips/boot/compressed/piggy.o
  LD      vmlinuz
ld.lld: error: section .text at 0xFFFFFFFF81000000 of size 0x3954 exceeds available address space
ld.lld: error: section .MIPS.abiflags at 0xFFFFFFFF81003958 of size 0x18 exceeds available address space
ld.lld: error: section .rodata.str1.1 at 0xFFFFFFFF81003970 of size 0x164 exceeds available address space
ld.lld: error: section .data at 0xFFFFFFFF81003AE0 of size 0x427270 exceeds available address space
ld.lld: error: section .got at 0xFFFFFFFF8142AD50 of size 0x8 exceeds available address space
ld.lld: error: section .appended_dtb at 0xFFFFFFFF8142AD58 of size 0x100000 exceeds available address space
ld.lld: error: section .bss at 0xFFFFFFFF8152AD60 of size 0x402010 exceeds available address space
...
```

This appears to be the same issue as #786 but for `zload-y` in `arch/mips/boot/compressed/Makefile`.

Commit [ff487d410360](https://git.kernel.org/linus/ff487d41036035376e47972c7c522490b839ab37) ("MIPS: Truncate link address into 32bit for 32bit kernel") took care of the case of no `zload-y` being defined by the platform (using `load-ld`/`LINKER_LOAD_ADDRESS`) but it did nothing for the case of `zload-y` being defined. As far as I can tell for this configuration, `zload-y` is set in `arch/mips/generic/Platform`:

```
$ rg MIPS_GENERIC .config
201:CONFIG_MIPS_GENERIC=y
207:# CONFIG_MIPS_GENERIC_KERNEL is not set
212:# CONFIG_BMIPS_GENERIC is not set

$ rg load- arch/mips/generic/Platform
15:load-$(CONFIG_MIPS_GENERIC)  += 0xffffffff80100000
16:zload-$(CONFIG_MIPS_GENERIC) += 0xffffffff81000000
```

This diff fixes it for me, no idea if this is proper or not.

```diff
diff --git a/arch/mips/boot/compressed/Makefile b/arch/mips/boot/compressed/Makefile
index f27cf31b4140..0d23acb1875c 100644
--- a/arch/mips/boot/compressed/Makefile
+++ b/arch/mips/boot/compressed/Makefile
@@ -90,8 +90,12 @@ HOSTCFLAGS_calc_vmlinuz_load_addr.o += $(LINUXINCLUDE)
 hostprogs := calc_vmlinuz_load_addr

 ifneq ($(zload-y),)
+ifdef CONFIG_64BIT
 VMLINUZ_LOAD_ADDRESS := $(zload-y)
 else
+VMLINUZ_LOAD_ADDRESS := $(subst 0xffffffff,0x,$(zload-y))
+endif
+else
 VMLINUZ_LOAD_ADDRESS = $(shell $(obj)/calc_vmlinuz_load_addr \
                $(obj)/vmlinux.bin $(LINKER_LOAD_ADDRESS))
 endif
```

cc @pcercuei

**Comment by pcercuei** [comment]
The diff looks fine to me.

**Comment by nathanchance** [comment]
This might be better:

```diff
diff --git a/arch/mips/boot/compressed/Makefile b/arch/mips/boot/compressed/Makefile
index f27cf31b4140..9d17ccb6bd8e 100644
--- a/arch/mips/boot/compressed/Makefile
+++ b/arch/mips/boot/compressed/Makefile
@@ -91,10 +91,16 @@ hostprogs := calc_vmlinuz_load_addr

 ifneq ($(zload-y),)
 VMLINUZ_LOAD_ADDRESS := $(zload-y)
+ifndef CONFIG_64BIT
+zload-ld := $(subst 0xffffffff,0x,$(zload-y))
+endif
 else
 VMLINUZ_LOAD_ADDRESS = $(shell $(obj)/calc_vmlinuz_load_addr \
                $(obj)/vmlinux.bin $(LINKER_LOAD_ADDRESS))
 endif
+ifeq ($(zload-ld),)
+zload-ld := $(VMLINUZ_LOAD_ADDRESS)
+endif
 UIMAGE_LOADADDR = $(VMLINUZ_LOAD_ADDRESS)

 vmlinuzobjs-y += $(obj)/piggy.o
@@ -102,7 +108,7 @@ vmlinuzobjs-y += $(obj)/piggy.o
 targets += ../../../../vmlinuz

 quiet_cmd_zld = LD      $@
-      cmd_zld = $(LD) $(KBUILD_LDFLAGS) -Ttext $(VMLINUZ_LOAD_ADDRESS) -T $< $(vmlinuzobjs-y) -o $@
+      cmd_zld = $(LD) $(KBUILD_LDFLAGS) -Ttext $(zload-ld) -T $< $(vmlinuzobjs-y) -o $@
 quiet_cmd_strip = STRIP   $@
       cmd_strip = $(STRIP) -s $@

```

Is it possible that this is related to #1333?

**Comment by pcercuei** [comment]
In this case why not replace `VMLINUZ_LOAD_ADDRESS` by `zload-ld` or vice-versa?

I thought about #1333 too, but it does not seem to be related - the vmlinuz file does have the proper zload address set.

**Comment by nathanchance** [comment]
It seems like `VMLINUZ_LOAD_ADDRESS ` has other uses? Looks like it is used for `uzImage.bin` (`UIMAGE_LOADADDR`) and `vmlinuz.its` (`VMLINUX_{LOAD,ENTRY}_ADDRESS`), I am not sure if the first change will impact those.

**Comment by JustinStitt** [comment]
Fixed in v5.8-rc1 with Commit ff487d4103603 ("MIPS: Truncate link address into 32bit for 32bit kernel")
